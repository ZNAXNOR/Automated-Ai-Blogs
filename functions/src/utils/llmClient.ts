
export class LLMClient {
  async generate(options: {
    model: string;
    prompt: string;
    max_tokens: number;
    temperature: number;
  }): Promise<{ text: string }> {
    // This is a placeholder and will be mocked in tests.
    return { text: "" };
  }
}
