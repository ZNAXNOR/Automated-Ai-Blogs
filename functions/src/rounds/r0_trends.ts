/* 
 * Deterministic, zero-cost by default. Optionally prunes/scores via a tiny HF model.
 *
 * Inputs:
 *   runId: string
 *   seeds: string[]
 *   region?: string
 *   useLLM?: boolean (default false)
 *
 * Side effects:
 *   - Writes artifact to Firestore: runs/{runId}.artifacts.round0
 *   - Caches SerpApi responses under cache/serpapi/{hash} with TTL
 *
 * Notes:
 *   - If SerpApi key not present, we fall back to RSS-only path.
 *   - The deterministic path is always executed; the LLM step is an optional extra prune.
 */

import * as admin from "firebase-admin";
import { HttpsError } from "firebase-functions/v2/https";
import { onCall } from "firebase-functions/v2/https";
import { FieldValue } from "firebase-admin/firestore";
import { env } from "../utils/config"; // expects SERP_API_KEY, HF_TOKEN, USE_R0_LLM, CACHE_TTL_HOURS
import { getSerpSuggestions, getSerpRelated, getSerpTrending, serpAvailable } from "../clients/serp";
import { getCache, setCache } from "../utils/cache";
import * as crypto from "crypto";

// --- Types -------------------------------------------------------------------

export interface TrendItem {
  query: string; // normalized short query
  type: "autocomplete" | "related" | "trending" | "rss";
  score: number; // 0..1
  source: string[]; // e.g., ["serp:autocomplete","rss:theverge"]
  reason?: string; // <= 12 words
}

interface Round0Input {
  runId: string;
  seeds: string[];
  region?: string;
  useLLM?: boolean;
  force?: boolean;
}

type SourceBucket = {
  type: TrendItem["type"];
  sourceName: string; // "serp:autocomplete" | "serp:related" | "serp:trending" | "rss:<domain>"
  items: string[]; // raw candidate strings (unnormalized)
};

// --- Admin init (safe if already initialized) --------------------------------
if (!admin.apps.length) {
  admin.initializeApp();
}
const db = admin.firestore();

// --- Lightweight RSS sources (fallback) --------------------------------------
// You can safely modify/extend this list.
const RSS_SOURCES: { name: string; url: string }[] = [
  { name: "rss:theverge", url: "https://www.theverge.com/rss/index.xml" },
  { name: "rss:techcrunch", url: "https://techcrunch.com/feed/" },
  { name: "rss:bbcworld", url: "https://feeds.bbci.co.uk/news/world/rss.xml" },
  { name: "rss:reutersworld", url: "https://feeds.reuters.com/reuters/worldNews" },
];

// --- Utilities ---------------------------------------------------------------

const ROUND = 0;

function logStep(runId: string, step: string, startMs: number) {
  const durationMs = Date.now() - startMs;
  // Use console.log for structured logs in emulator & prod
  console.log(JSON.stringify({ runId, round: ROUND, step, durationMs }));
}

function assertInput(i: any): asserts i is Round0Input {
  if (!i || typeof i !== "object") throw new HttpsError("invalid-argument", "Input must be an object");
  if (!i.runId || typeof i.runId !== "string") throw new HttpsError("invalid-argument", "runId is required");
  if (!Array.isArray(i.seeds) || i.seeds.some((s: string) => typeof s !== "string")) {
    throw new HttpsError("invalid-argument", "seeds must be a string[]");
  }
  if (i.region && typeof i.region !== "string") {
    throw new HttpsError("invalid-argument", "region must be a string");
  }
  if (i.useLLM !== undefined && typeof i.useLLM !== "boolean") {
    throw new HttpsError("invalid-argument", "useLLM must be boolean");
  }
}

function sha256(str: string) {
  return crypto.createHash("sha256").update(str, "utf8").digest("hex");
}

// Keep '+', '#', and '-'; strip other punctuation.
function stripPunctKeepSymbols(s: string): string {
  return s.replace(/[!"$%&'()*.,\/:;<=>?@\[\\\]^_`{|}~]/g, " "); // keep +, #, and -
}

// Normalize: lowercase, trim, collapse spaces, truncate to 6 words
export function normalizeQuery(raw: string): string {
  const s = stripPunctKeepSymbols(raw.toLowerCase())
    .replace(/\s+/g, " ")
    .trim();
  const words = s.split(" ").filter(Boolean).slice(0, 6);
  return words.join(" ");
}

// Filters per spec
export function shouldDrop(q: string): boolean {
  const tokens = q.split(" ").filter(Boolean);
  const hasPersonal = /\b(my|me|account|password)\b/.test(q);
  if (hasPersonal) return true;

  const numericCount = tokens.filter(t => /^\d+([\/\.\-]\d+)*$/.test(t)).length;
  if (tokens.length > 0 && numericCount / tokens.length > 0.6) return true;

  const generic = ["news", "update", "latest"];
  const containsGeneric = tokens.some(t => generic.includes(t));
  if (containsGeneric) {
    // allow if paired with context like "apple latest" (â‰¥ 2 tokens and not only generic)
    if (tokens.length === 1 && generic.includes(tokens[0])) return true;
    const nonGenericCount = tokens.filter(t => !generic.includes(t)).length;
    if (nonGenericCount === 0) return true;
  }
  return false;
}

// Simple Levenshtein (enough for small N)
function levenshtein(a: string, b: string): number {
  if (a === b) return 0;
  if (a.length === 0) return b.length;
  if (b.length === 0) return a.length;
  const v0 = Array(b.length + 1).fill(0);
  const v1 = Array(b.length + 1).fill(0);
  for (let i = 0; i <= b.length; i++) v0[i] = i;
  for (let i = 0; i < a.length; i++) {
    v1[0] = i + 1;
    for (let j = 0; j < b.length; j++) {
      const cost = a[i] === b[j] ? 0 : 1;
      v1[j + 1] = Math.min(v1[j] + 1, v0[j + 1] + 1, v0[j] + cost);
    }
    for (let j = 0; j <= b.length; j++) v0[j] = v1[j];
  }
  return v1[b.length];
}

function tokenOverlap(a: string, b: string): number {
  const A = new Set(a.split(" ").filter(Boolean));
  const B = new Set(b.split(" ").filter(Boolean));
  if (A.size === 0 && B.size === 0) return 1;
  let inter = 0;
  A.forEach(t => {
    if (B.has(t)) inter++;
  });
  const union = new Set([...A, ...B]).size;
  return inter / union;
}

function nearDuplicate(a: string, b: string): boolean {
  return levenshtein(a, b) <= 2 || tokenOverlap(a, b) >= 0.75;
}

// Optionally call a tiny HF LLM to prune/score (kept very cheap)
// If env.HF_TOKEN missing or useLLM=false, skip and return passthrough.
async function optionalLlmPrune(
  items: TrendItem[],
  useLLM: boolean | undefined
): Promise<TrendItem[]> {
  if (!useLLM && !env.useR0Llm) return items;
  if (!env.HF_TOKEN) return items;

  // Token-efficient prompt: rank by general audience interest + freshness proxies
  const prompt =
`You are scoring short search queries for blog topics.
Keep 12 or fewer. Prefer widely interesting, multi-source, and non-duplicative.
Return as CSV: query,boost where boost in [0..0.2].

Items:
${items.map(i => `- ${i.query} [${i.type}]`).join("\n")}
`;

  try {
    // Lazy import to avoid dependency if unused
    const { hfTinyComplete } = await import("../clients/hf");
    const csv = await hfTinyComplete(prompt); // returns small CSV string or empty
    if (!csv) return items;

    const boosts = new Map<string, number>();
    csv.split(/\r?\n/).forEach(line => {
      const m = line.split(",");
      if (m.length >= 2) {
        const q = normalizeQuery(m[0] || "");
        const b = Number(m[1]);
        if (q && isFinite(b)) boosts.set(q, Math.max(0, Math.min(0.2, b)));
      }
    });

    const boosted = items.map(i => ({
      ...i,
      score: Math.max(0, Math.min(1, i.score + (boosts.get(i.query) ?? 0))),
    }));

    // Re-sort and clamp to 12
    boosted.sort((a, b) => (b.score - a.score) || (a.query.length - b.query.length));
    return boosted.slice(0, 12);
  } catch (e) {
    console.warn("LLM prune skipped due to error:", e);
    return items;
  }
}

// --- RSS fetch & parse (lightweight) ----------------------------------------
// Minimal parser: extract <title>...</title> from top items (ignores CDATA nuances gracefully)
async function fetchRssTitles(url: string, limit = 20): Promise<string[]> {
  const res = await fetch(url, { method: "GET", redirect: "follow" });
  const text = await res.text();
  const titles: string[] = [];
  const re = /<title>([\s\S]*?)<\/title>/gi;
  let m: RegExpExecArray | null;
  while ((m = re.exec(text)) !== null) {
    const raw = m[1]
      .replace(/<!\[CDATA\[(.*?)\]\]>/g, "$1")
      .replace(/\s+/g, " ")
      .trim();
    // Skip the feed-level title (usually first one)
    if (raw && titles.length < limit) titles.push(raw);
  }
  // Remove the first title if it is the channel title (heuristic)
  return titles.length > 1 ? titles.slice(1) : titles;
}

// --- Core deterministic pipeline --------------------------------------------

export function deterministicProcess(
  buckets: SourceBucket[]
): { items: TrendItem[]; sourceCounts: Record<string, number> } {
  const collected = new Map<string, TrendItem>();

  for (const b of buckets) {
    for (const raw of b.items) {
      const q = normalizeQuery(raw);
      if (!q || shouldDrop(q)) continue;

      if (collected.has(q)) {
        // Merge source/type
        const existing = collected.get(q)!;
        const newSources = new Set([...existing.source, b.sourceName]);
        existing.source = [...newSources];
        const typeRank = (t: TrendItem["type"]) =>
          t === "trending" ? 3 : t === "related" ? 2 : t === "autocomplete" ? 1 : 0;
        if (typeRank(b.type) > typeRank(existing.type)) {
          existing.type = b.type;
        }
      } else {
        collected.set(q, {
          query: q,
          type: b.type,
          score: 0,
          source: [b.sourceName],
        });
      }
    }
  }

  // Near-duplicate merge
  const merged: TrendItem[] = [];
  for (const item of Array.from(collected.values())) {
    const dupIdx = merged.findIndex(m => nearDuplicate(m.query, item.query));
    if (dupIdx === -1) {
      merged.push(item);
    } else {
      // merge sources and keep best type (prefer trending > related > autocomplete > rss)
      const target = merged[dupIdx];
      const src = new Set([...target.source, ...item.source]);
      target.source = [...src];
      const typeRank = (t: TrendItem["type"]) =>
        t === "trending" ? 3 : t === "related" ? 2 : t === "autocomplete" ? 1 : 0;
      if (typeRank(item.type) > typeRank(target.type)) target.type = item.type;
    }
  }

  // Scoring
  const fromTrending = new Set(
    merged.filter(i => i.source.some(s => s.startsWith("serp:trending"))).map(i => i.query)
  );

  const itemsScored = merged.map(i => {
    let score = 0.5; // base
    const multiSource = i.source.length > 1;
    if (multiSource) score += 0.10;
    if (fromTrending.has(i.query)) score += 0.10;
    score = Math.max(0, Math.min(1, score));
    const reasonBits = [];
    if (fromTrending.has(i.query)) reasonBits.push("trending");
    if (multiSource) reasonBits.push("multi-source");
    const reason = reasonBits.length ? reasonBits.join(", ").slice(0, 48) : undefined;
    return { ...i, score, reason };
  });

  // Sort by score desc then by query length asc; clamp 12
  itemsScored.sort((a, b) => (b.score - a.score) || (a.query.length - b.query.length));
  const final = itemsScored.slice(0, 12);

  // sourceCounts
  const sourceCounts: Record<string, number> = {};
  for (const i of final) {
    for (const s of i.source) sourceCounts[s] = (sourceCounts[s] ?? 0) + 1;
  }

  return { items: final, sourceCounts };
}

// --- Firestore artifact helpers ---------------------------------------------

async function getExistingArtifact(runId: string) {
  const ref = db.doc(`runs/${runId}`);
  const snap = await ref.get();
  const data = snap.exists ? snap.data() : undefined;
  const art = data?.artifacts?.round0;
  return art ? { ...art } : null;
}

async function writeArtifact(runId: string, payload: any) {
  const ref = db.doc(`runs/${runId}`);
  await ref.set(
    {
      artifacts: {
        round0: {
          ...payload,
          createdAt: FieldValue.serverTimestamp(),
        },
      },
    },
    { merge: true }
  );
}

// --- Main callable -----------------------------------------------------------

export const round0_trends = onCall(
  { timeoutSeconds: 300, memory: "256MiB" },
  async (req) => {
    const t0 = Date.now();
    const input = req.data;
    assertInput(input);
    const { runId, seeds, region, useLLM, force } = input;

    // Idempotency
    if (!force) {
      const existing = await getExistingArtifact(runId);
      if (existing) {
        logStep(runId, "idempotent-return", t0);
        return existing;
      }
    }

    const tSerp = Date.now();

    // --- Gather sources (Serp + RSS fallback) --------------------------------
    const buckets: SourceBucket[] = [];

    // SerpApi (if available)
    let usedSerp = false;
    let serpCacheHit = false;

    if (serpAvailable()) {
      usedSerp = true;
      const dayBucket = new Date().toISOString().slice(0, 10); // YYYY-MM-DD for cache key
      const baseKey = JSON.stringify({ seeds, region, day: dayBucket });

      // Autocomplete
      {
        const cacheKey = `serpapi:${sha256(baseKey + ":autocomplete")}`;
        const cached = await getCache(cacheKey);
        let list: string[];
        if (cached?.payload) {
          list = cached.payload;
          serpCacheHit = true;
        } else {
          list = await getSerpSuggestions(seeds, region);
          await setCache(cacheKey, list, env.CACHE_TTL_HOURS);
        }
        buckets.push({ type: "autocomplete", sourceName: "serp:autocomplete", items: list });
      }
      // Related
      {
        const cacheKey = `serpapi:${sha256(baseKey + ":related")}`;
        const cached = await getCache(cacheKey);
        let list: string[];
        if (cached?.payload) {
          list = cached.payload;
          serpCacheHit = true;
        } else {
          list = await getSerpRelated(seeds, region);
          await setCache(cacheKey, list, env.CACHE_TTL_HOURS);
        }
        buckets.push({ type: "related", sourceName: "serp:related", items: list });
      }
      // Trending
      {
        const cacheKey = `serpapi:${sha256(baseKey + ":trending")}`;
        const cached = await getCache(cacheKey);
        let list: string[];
        if (cached?.payload) {
          list = cached.payload;
          serpCacheHit = true;
        } else {
          list = await getSerpTrending(region);
          await setCache(cacheKey, list, env.CACHE_TTL_HOURS);
        }
        buckets.push({ type: "trending", sourceName: "serp:trending", items: list });
      }
    }

    // RSS fallback (always included, but especially useful if Serp is unavailable)
    const tRss = Date.now();
    for (const src of RSS_SOURCES) {
      try {
        const titles = await fetchRssTitles(src.url, 20);
        buckets.push({ type: "rss", sourceName: src.name, items: titles });
      } catch (e) {
        console.warn(`RSS fetch error for ${src.name}`, e);
      }
    }
    logStep(runId, "fetch-sources", tSerp);

    // --- Deterministic normalization/dedup/score ------------------------------
    const tDet = Date.now();
    const { items: detItems, sourceCounts } = deterministicProcess(buckets);
    logStep(runId, "deterministic", tDet);

    // --- Optional LLM prune/boost --------------------------------------------
    const tLlm = Date.now();
    const finalItems = await optionalLlmPrune(detItems, useLLM);
    logStep(runId, "optional-llm", tLlm);

    // --- Output validation (simple) ------------------------------------------
    for (const i of finalItems) {
      if (!i.query || typeof i.query !== "string") throw new HttpsError("internal", "Invalid item.query");
      if (!["autocomplete", "related", "trending", "rss"].includes(i.type)) {
        throw new HttpsError("internal", "Invalid item.type");
      }
      if (typeof i.score !== "number" || i.score < 0 || i.score > 1) {
        throw new HttpsError("internal", "Invalid item.score");
      }
      if (!Array.isArray(i.source)) throw new HttpsError("internal", "Invalid item.source");
      if (i.reason && i.reason.split(/\s+/).length > 12) {
        throw new HttpsError("internal", "Invalid item.reason length");
      }
    }

    // --- Write artifact -------------------------------------------------------
    const artifact = {
      items: finalItems,
      cached: serpCacheHit,
      sourceCounts,
    };
    await writeArtifact(runId, artifact);

    logStep(runId, "done", t0);
    return artifact;
  }
);

// Export internals for unit testing
export const _test = {
  normalizeQuery,
  shouldDrop,
  deterministicProcess,
  nearDuplicate,
};
