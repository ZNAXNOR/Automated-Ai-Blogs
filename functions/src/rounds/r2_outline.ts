/**
 * Reads R1 artifact (ideation), generates structured outlines per idea
 * using Hugging Face, and saves results back to Firestore.
 */

import { getFirestore } from "firebase-admin/firestore";
import fetch from "node-fetch";
import { env } from "../utils/config";

// ---- Types ----

export interface IdeationItem {
  trend: string;
  idea: string;
  variant: number;
  source: "llm";
}

export interface OutlineItem {
  trend: string;
  idea: string;
  sections: {
    heading: string;
    bullets: string[];
    estWordCount: number;
  }[];
}

// ---- Prompt Template ----

function buildPrompt(ideas: IdeationItem[]): string {
  return `
You are a professional blog content planner.

TASK:
- Create a compact blog post outline for each input idea.
- Include 3–6 sections.
- Each section should have:
  - A clear heading
  - 2–3 bullet points
  - An estimated word count (50–150 words)
- Keep outlines logical, engaging, and suitable for SEO.

OUTPUT FORMAT (strict JSON only):
[
  {
    "trend": "<trend string>",
    "idea": "<headline string>",
    "sections": [
      {
        "heading": "<section heading>",
        "bullets": ["<point1>", "<point2>", "<point3>"],
        "estWordCount": 120
      }
    ]
  }
]

Input Ideas:
${JSON.stringify(ideas, null, 2)}
  `.trim();
}

// ---- Model Call ----

async function callHuggingFace(prompt: string): Promise<OutlineItem[]> {
  const HF_API_URL = `https://api-inference.huggingface.co/models/${env.hfModelR2}`;

  const response = await fetch(HF_API_URL, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${env.HF_API_KEY}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      inputs: prompt,
      parameters: {
        max_new_tokens: 800,
        temperature: 0.7,
      },
    }),
  });

  if (!response.ok) {
    throw new Error(`HF API error: ${response.status} ${response.statusText}`);
  }

  const data = (await response.json()) as any;

  let text: string;
  if (Array.isArray(data) && data.length > 0 && data[0].generated_text) {
    text = data[0].generated_text;
  } else if (typeof data === "object" && data.generated_text) {
    text = data.generated_text;
  } else {
    text = JSON.stringify(data);
  }

  try {
    const parsed = JSON.parse(text);
    return parsed as OutlineItem[];
  } catch (err) {
    console.error("Failed to parse JSON from model:", text);
    throw err;
  }
}

// ---- Main Runner ----

export async function runRound2Outline(runId: string): Promise<OutlineItem[]> {
  const db = getFirestore();

  // 1. Read ideation artifact (Round 1)
  const r1Snap = await db
    .doc(`runs/${runId}/artifacts/round1`)
    .get();
  if (!r1Snap.exists) {
    throw new Error(`Round1 artifact not found for runId=${runId}`);
  }

  const r1Data = r1Snap.data() || {};
  const items: IdeationItem[] = r1Data.items || [];
  if (items.length === 0) {
    throw new Error(`No ideas found in Round1 artifact for runId=${runId}`);
  }

  // 2. Build prompt
  const prompt = buildPrompt(items);

  // 3. Call Model
  const outlines = await callHuggingFace(prompt);

  // 4. Save to Firestore
  await db
    .doc(`runs/${runId}/artifacts/round2`)
    .set({ items: outlines }, { merge: true });

  return outlines;
}

// Expose helpers for testing
export const _test = {
  buildPrompt,
  callHuggingFace,
};
