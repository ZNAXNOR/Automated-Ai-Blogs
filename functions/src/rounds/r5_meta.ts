/**
 * Implements metadata generation for polished drafts.
 * - Reads drafts from an array of strings.
 * - Generates SEO metadata and image suggestions via GenKit/LLM.
 * - Validates outputs and returns an array of metadata objects.
 */

import { getFirestore, FieldValue } from "firebase-admin/firestore";
import { LLMClient } from "../utils/llmClient";

// Constraints
const MAX_TITLE = 70;
const MAX_DESC = 160;
const EXCERPT_MIN_WORDS = 50;
const EXCERPT_MAX_WORDS = 100;
const MIN_DRAFT_LENGTH = 250; // Minimum characters for a draft to be processed

function countWords(text: string): number {
  return text.trim().split(/\s+/).filter(Boolean).length;
}

interface Metadata {
  seoTitle: string;
  metaDescription: string;
  tags: string[];
  categories: string[];
  excerpt: string;
  relatedKeywords: string[];
  imageSuggestions: string[];
}

function validateMetadata(parsed: any): Metadata {
  if (!parsed.seoTitle || parsed.seoTitle.length > MAX_TITLE) {
    throw new Error("Invalid seoTitle");
  }
  if (!parsed.metaDescription || parsed.metaDescription.length > MAX_DESC) {
    throw new Error("Invalid metaDescription");
  }
  if (
    !parsed.excerpt ||
    countWords(parsed.excerpt) < EXCERPT_MIN_WORDS ||
    countWords(parsed.excerpt) > EXCERPT_MAX_WORDS
  ) {
    throw new Error("Invalid excerpt");
  }
  if (!Array.isArray(parsed.tags) || parsed.tags.length < 3) {
    throw new Error("Invalid tags");
  }
  if (!Array.isArray(parsed.categories) || parsed.categories.length < 1) {
    throw new Error("Invalid categories");
  }
  if (
    !Array.isArray(parsed.relatedKeywords) ||
    parsed.relatedKeywords.length < 3
  ) {
    throw new Error("Invalid relatedKeywords");
  }
  if (
    !Array.isArray(parsed.imageSuggestions) ||
    parsed.imageSuggestions.length === 0
  ) {
    throw new Error("Invalid imageSuggestions");
  }
  return parsed as Metadata;
}

async function generateMetaForDraft(draft: string, llm: LLMClient): Promise<Metadata | null> {
  if (!draft || draft.trim().length < MIN_DRAFT_LENGTH) {
    console.warn("R5: Skipping draft, too short.");
    return null;
  }

  const prompt = `
    System: You are an expert SEO strategist. Generate the following outputs as a single, valid JSON object:
    1. seoTitle (string, <= ${MAX_TITLE} chars)
    2. metaDescription (string, <= ${MAX_DESC} chars)
    3. tags (array of strings, min 3)
    4. categories (array of strings, min 1)
    5. excerpt (string, ${EXCERPT_MIN_WORDS}-${EXCERPT_MAX_WORDS} words)
    6. relatedKeywords (array of strings, min 3)
    7. imageSuggestions (array of strings, at least one prompt or reuse)

    Draft:
    ${draft}
  `;

  try {
    const res = await llm.generate({
      model: "gemini-1.5-flash",
      prompt,
      max_tokens: 800,
      temperature: 0.1,
    });

    const jsonText = res.text.match(/\{[\s\S]*\}/)?.[0];
    if (!jsonText) throw new Error("No valid JSON found in LLM response.");

    const parsed = JSON.parse(jsonText);
    return validateMetadata(parsed);

  } catch (error) {
    console.error("Error generating metadata:", error);
    return null; 
  }
}

export async function Round5_Meta(runId: string): Promise<void> {
    const db = getFirestore();
    const llm = new LLMClient();
    console.log(`R5: Starting Round 5 for runId=${runId}`);
  
    const r4ArtifactRef = db.doc(`runs/${runId}/artifacts/round4`);
    const r4ArtifactSnapshot = await r4ArtifactRef.get();
  
    if (!r4ArtifactSnapshot.exists) {
      console.log(`R5: No Round 4 artifact found. Exiting.`);
      return;
    }
    
    const r4Data = r4ArtifactSnapshot.data();
    const draftsToProcess = r4Data?.items || [];

    console.log(`R5: Found ${draftsToProcess.length} drafts to process.`);
  
    const metadataPromises = draftsToProcess.map(async (draft: any) => {
      const metadata = await generateMetaForDraft(draft.polished, llm);
      return { draftId: draft.id, metadata };
    });
  
    const results = await Promise.all(metadataPromises);
    const successfulResults = results.filter((r) => r.metadata);
    const failedCount = results.length - successfulResults.length;
  
    console.log(`R5: Successfully generated metadata for ${successfulResults.length} drafts.`);
    if (failedCount > 0) {
      console.warn(`R5: Failed to generate metadata for ${failedCount} drafts.`);
    }
  
    if (successfulResults.length > 0) {
      const batch = db.batch();
      const r5Collection = db.collection(`runs/${runId}/artifacts/round5_distribution`);
  
      successfulResults.forEach(({ draftId, metadata }) => {
        const docRef = r5Collection.doc(draftId);
        batch.set(docRef, {
          ...metadata,
          createdAt: FieldValue.serverTimestamp(),
        });
      });
  
      await batch.commit();
      console.log(`R5: Saved ${successfulResults.length} metadata artifacts.`);
    }
  
    await db.doc(`runs/${runId}/artifacts/round5`).set({
      status: "completed",
      successCount: successfulResults.length,
      failureCount: failedCount,
      updatedAt: FieldValue.serverTimestamp(),
    });
  
    console.log(`R5: Round 5 finished for runId=${runId}.`);
  }

export const _test = {
    generateMetaForDraft
};
