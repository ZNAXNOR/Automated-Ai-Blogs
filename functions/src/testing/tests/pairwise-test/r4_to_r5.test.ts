jest.mock("../../../utils/llmClient");

import { _test as r5_test_functions } from "../../../rounds/r5_meta";
const { generateMetaForDraft } = r5_test_functions;

import { LLMClient } from "../../../utils/llmClient";

const llmMock = new LLMClient() as jest.Mocked<LLMClient>;

const r4Sample = {
  id: "test-draft-1",
  polished: `
    The future of AI in marketing is not just about automation; it's about creating hyper-personalized experiences that resonate with customers on a deeper level. This draft explores how AI can analyze vast amounts of customer data to predict behavior, tailor content, and optimize campaigns in real-time. We delve into the ethical considerations and the importance of transparency in AI-driven marketing, ensuring that customer trust is maintained. The article also provides actionable insights for marketers looking to adopt AI tools, highlighting the potential for unprecedented efficiency and engagement. By leveraging AI, brands can build deeper, more meaningful relationships with their audience, turning data into a competitive advantage. This is not a distant future; it's happening now, and early adopters are already reaping the rewards. We examine case studies from leading brands that have successfully integrated AI into their marketing strategies. From chatbots that provide instant support to predictive analytics that forecast trends, the applications are vast. The key is to balance technological power with a human-centric approach, ensuring that AI serves rather than dictates. This comprehensive guide will equip you with the knowledge to navigate the evolving landscape of AI in marketing.
  `,
  derivatives: [
    "AI is revolutionizing marketing by enabling hyper-personalized content. #AI #Marketing",
    "Leverage AI to predict customer behavior and optimize your marketing campaigns in real-time.",
  ],
};

const validMockResponse = {
    seoTitle: "AI Marketing Trends: 2025 Guide",
    metaDescription: "Learn how AI is reshaping content creation and marketing workflows.",
    tags: ["AI", "Marketing", "Content Automation"],
    categories: ["Marketing", "Technology"],
    excerpt: "This is a detailed exploration into how artificial intelligence is fundamentally changing the marketing landscape. We cover everything from personalized content creation and automated campaign management to the ethical challenges that arise. Our guide provides practical, actionable advice for businesses looking to integrate AI into their workflows, ensuring they can leverage these powerful tools to connect with audiences more effectively and achieve a higher return on investment in a competitive digital world.",
    relatedKeywords: ["AI in marketing", "content automation", "AI marketing tools"],
    imageSuggestions: ["prompt: AI robot creating content", "reuse: company logo"]
};


describe("R4 â†’ R5 Pairwise Metadata Generation", () => {
  beforeEach(() => {
    llmMock.generate.mockClear();
    llmMock.generate.mockResolvedValue({ text: JSON.stringify(validMockResponse) });
  });

  it("generates metadata for a single polished draft and matches snapshot", async () => {
    const result = await Promise.all([r4Sample.polished].map(draft => generateMetaForDraft(draft, llmMock)));
    expect(result.length).toBe(1);
    const meta = result[0]!;
    expect(meta.seoTitle).toBeDefined();
    expect(meta.tags.length).toBeGreaterThan(0);
    expect(meta).toMatchSnapshot();
  });

  it("generates metadata for multiple R4 items", async () => {
    const items = [
      r4Sample.polished,
      r4Sample.polished, // Using the same long draft twice
      r4Sample.polished
    ];
    const result = await Promise.all(items.map(draft => generateMetaForDraft(draft, llmMock)));
    expect(result.length).toBe(items.length);
    // Ensure LLM was called once per item
    expect(llmMock.generate).toHaveBeenCalledTimes(items.length);
  });

  it("ensures metadata fields are non-empty and arrays are valid", async () => {
    const result = await Promise.all([r4Sample.polished].map(draft => generateMetaForDraft(draft, llmMock)));
    const meta = result[0]!;
    expect(meta.seoTitle).toBeTruthy();
    expect(meta.metaDescription).toBeTruthy();
    expect(meta.excerpt).toBeTruthy();
    expect(meta.tags.length).toBeGreaterThan(0);
    expect(meta.categories.length).toBeGreaterThan(0);
    expect(meta.relatedKeywords.length).toBeGreaterThan(0);
    expect(meta.imageSuggestions.length).toBeGreaterThan(0);
  });

  it("validates content length limits", async () => {
    const result = await Promise.all([r4Sample.polished].map(draft => generateMetaForDraft(draft, llmMock)));
    const meta = result[0]!;
    expect(meta.seoTitle.length).toBeLessThanOrEqual(70);
    expect(meta.metaDescription.length).toBeLessThanOrEqual(160);
    const wordCount = meta.excerpt.split(/\s+/).filter(Boolean).length;
    expect(wordCount).toBeGreaterThanOrEqual(50);
    expect(wordCount).toBeLessThanOrEqual(100);
  });
});
