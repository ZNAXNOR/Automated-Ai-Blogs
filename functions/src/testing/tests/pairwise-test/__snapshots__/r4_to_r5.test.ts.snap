// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`R4 â†’ R5 Pairwise Metadata Generation generates metadata for a single polished draft and matches snapshot 1`] = `
{
  "categories": [
    "Marketing",
    "Technology",
  ],
  "excerpt": "This is a detailed exploration into how artificial intelligence is fundamentally changing the marketing landscape. We cover everything from personalized content creation and automated campaign management to the ethical challenges that arise. Our guide provides practical, actionable advice for businesses looking to integrate AI into their workflows, ensuring they can leverage these powerful tools to connect with audiences more effectively and achieve a higher return on investment in a competitive digital world.",
  "imageSuggestions": [
    "prompt: AI robot creating content",
    "reuse: company logo",
  ],
  "metaDescription": "Learn how AI is reshaping content creation and marketing workflows.",
  "relatedKeywords": [
    "AI in marketing",
    "content automation",
    "AI marketing tools",
  ],
  "seoTitle": "AI Marketing Trends: 2025 Guide",
  "tags": [
    "AI",
    "Marketing",
    "Content Automation",
  ],
}
`;
